bpf/bpf_overlay.c

```diff
@@ -321,16 +323,18 @@ static __always_inline int handle_ipv4(struct __ctx_buff *ctx,
 	decrypted = ((ctx->mark & MARK_MAGIC_HOST_MASK) == MARK_MAGIC_DECRYPT);
 	/* If packets are decrypted the key has already been pushed into metadata. */
 	if (decrypted) {
-		if (info)
-			*identity = key.tunnel_id = info->sec_identity;
+		if (info) {
+			*identity = info->sec_identity;
+			key.tunnel_id = get_tunnel_id(info->sec_identity);
+		}
 	} else {
 #ifdef ENABLE_HIGH_SCALE_IPCACHE
-		key.tunnel_id = *identity;
+		key.tunnel_id = get_tunnel_id(*identity);
```

Comment:
This should also be `get_id_from_tunnel_id()` right? Effectively that function is the "receive" side converter, while `get_tunnel_id()` is the transmit side?
```suggestion
		key.tunnel_id = get_id_from_tunnel_id(*identity);
```

Reply:
No. Any tunnel key look-up should always convert `WORLD_IPVX_ID` to `WORLD_ID`. It doesn't matter if its on the receive-side.

Reply:
I disagree. What's important is that on the wire, `WORLD_ID` is used to ensure compatibility with older releases. Internally in the datapath if running in dual-stack mode, the datapath should consistently identify the identity of that peer and if it's an IPv4-based world identity peer that should be `WORLD_IPV4_ID`. That way, when we receive traffic with `WORLD_ID`, we can convert that to `WORLD_IPV4_ID`, then generate hubble flows with `WORLD_IPV4_ID` and do network policy lookups with `WORLD_IPV4_ID` in order to properly reflect that this new version of Cilium is identifying this traffic as a more narrowly scoped "world" for the specific protocol.

At the minimum, the policy lookup needs to use the protocol-scoped world identity to perform the lookup, otherwise the traffic could be denied by an "allow `0.0.0.0/0`" policy. If we're doing that for policy lookup, then the datapath should likewise emit all hubble events with the identity that the datapath is considering this traffic to correspond to. Otherwise we'll have a lot of difficulty debugging problems related to these new protocol-specific world identities.

Reply:
Reading through this afresh, I assumed that with the `#ifdef ENABLE_HIGH_SCALE_IPCACHE` function that the macro if/else was updating the same variables / fields and that this line was updating `*identity`. I then argued above that internally within the datapath we should always be using the newer protocol-specific world identities. The first assumption was wrong and this code is enforcing the exact property that I was arguing for. :sweat_smile: 

Sorry for the confusion! LGTM.

Reply:
(I'm now a bit confused why the high scale ipcache code here doesn't update `*identity`, but that has nothing to do with this PR).